src/openslide-vendor-dicom.c

```diff
@@ -0,0 +1,932 @@
+                                 gpointer value, 
+                                 gpointer user_data G_GNUC_UNUSED) {
+  struct dicom_file *f = (struct dicom_file *) value;
+
+  print_file(f);
+}
+
+static void print_level_from_hash(gpointer key G_GNUC_UNUSED, 
+                                  gpointer value, 
+                                  gpointer user_data G_GNUC_UNUSED) {
+  struct dicom_level *l = (struct dicom_level *) value;
+
+  print_level(l);
+}
+
+static int find_associated(gpointer key G_GNUC_UNUSED, 
+                            gpointer value, 
+                            gpointer user_data) {
+  struct dicom_file *f = (struct dicom_file *) value;
+  openslide_t *osr = (openslide_t *) user_data; 
+
+  struct associated *a = associated_new(f);
+  if (a) {
+    // map dicom associated image names to openslide names
+    char *openslide_name = NULL;
+    if (g_str_equal(a->name, "LABEL")) {
+      openslide_name = "label";
+    }
+    else if (g_str_equal(a->name, "OVERVIEW")) {
+      openslide_name = "macro";
+    }
+
+    if (openslide_name) {
+      g_hash_table_insert(osr->associated_images, 
+                          g_strdup(openslide_name), 
+                          a);
+      return true;
+    }
+  }
+
+  return false;
+}
+
+static gint compare_level_width(const void *a, const void *b) {
+  const struct dicom_level *aa = *((const struct dicom_level **) a);
+  const struct dicom_level *bb = *((const struct dicom_level **) b);
+
+  return bb->image_width - aa->image_width;
+}
+
+static void make_grid(gpointer key G_GNUC_UNUSED, 
+                      gpointer value, 
+                      gpointer user_data) {
+  struct dicom_level *l = (struct dicom_level *) value;
+  openslide_t *osr = (openslide_t *) user_data; 
+
+  l->grid = _openslide_grid_create_simple(osr,
+                                          l->tiles_across, 
+                                          l->tiles_down,
+                                          l->tile_w, 
+                                          l->tile_h,
+                                          read_tile);
+}
+
+static void add_properties(openslide_t *osr, const struct dicom_level *l) {
+  // why not
+  double mmpp = l->image_width;
+  g_hash_table_insert(osr->properties,
+                      g_strdup(OPENSLIDE_PROPERTY_NAME_MPP_X),
+                      _openslide_format_double(mmpp * 1000));
+  g_hash_table_insert(osr->properties,
+                      g_strdup(OPENSLIDE_PROPERTY_NAME_MPP_Y),
+                      _openslide_format_double(mmpp * 1000));
+}
+
+static void file_hash(const struct dicom_file *f,
+                      struct _openslide_hash *quickhash1) {
+  uint32_t n = dcm_dataset_count(f->metadata);
+  for(uint32_t i = 0; i < n; i++) {
+    DcmElement *element = dcm_dataset_get(NULL, f->metadata, i);
+    if (element) {
+      uint32_t tag = dcm_element_get_tag(element);
+      _openslide_hash_data(quickhash1, &tag, sizeof(tag));
```

Comment:
Could you spell out exactly what this is doing?  We need to be careful about the definition of the quickhash since it needs to be stable, different for every slide, and fast.  We usually put the quickhash definition in a comment at the top of the source file.

Note that we often hash in the image data from the smallest slide level.

Reply:
This is a placeholder. I'm planning to fix this up in phase 2 when we get more complete metadata support.

Reply:
We shouldn't change an existing quickhash once it's been introduced.  For now, we can call `_openslide_hash_disable()`.